Store TempDir for the lifetime of a program

blackbeam opened the pull request:
This solves #34.

Might be an overkill but i can't find another solution.
I've tried to store tempdir within Pkcs12 and even tried to store it in a wrapper for SecKeychain but with no luck.

sfackler commented on src/imp/security_framework.rs:
```diff
@@ -14,9 +15,16 @@ use self::tempdir::TempDir;
 
+lazy_static! {
+    static ref AT_EXIT_SET: AtomicBool = AtomicBool::new(false);
+    static ref TEMP_DIRS: Mutex<Vec<TempDir>> = Mutex::new(vec![]);
```
Can we treat this as a map from the raw DER data to the directory? That way we don't end up with tons of directories when the same identity is imported multiple times.

sfackler commented on src/imp/security_framework.rs:
```diff
@@ -101,6 +110,16 @@ impl Pkcs12 {
         // FIXME: Compare the certificates for equality using CFEqual
         let identity_cert = try!(import.identity.certificate()).to_der();
 
+        if !AT_EXIT_SET.swap(true, Ordering::Relaxed) {
```
You can use std::sync::Once to do this.

blackbeam commented:
If SecKeystore isn't cached, then you'll get "keychain with this name already exists" error, so now it's also cached.

sfackler commented on src/imp/security_framework.rs:
```diff
@@ -79,17 +94,27 @@ pub struct Pkcs12 {
-        // disable lock on sleep and timeouts
-        try!(keychain.set_settings(&KeychainSettings::new()));
-
         let mut imports = try!(
```
Will things behave okay if we import the same identity into the same keychain twice?

blackbeam commented:
>Why do we need a custom hash function here?

It's faster for long keys. It's just a habit and i'll remove it if you say so.

> Will things behave okay if we import the same identity into the same keychain twice?

I've tried to import same der multiple times and it looks like the number of identities in a keychain stays the same (1 identity as shown by ItemSearchOptions).
Anyway there is no need to import same der multiple times so cache now simply stores `Pkcs12` instance.

sfackler commented:
I'd like to see some benchmarks indicating that the custom hasher is worth it.

The new behavior does mean the password is ignored after the first decode for a given identity.

blackbeam commented:
> I'd like to see some benchmarks indicating that the custom hasher is worth it.

[this one](http://aras-p.info/blog/2016/08/02/Hash-Functions-all-the-way-down/)

>The new behavior does mean the password is ignored after the first decode for a given identity.

🤦‍♂️
Well. First of all i tried to simply import pkcs12 without specifying keychain but than i found commit message of 304b00e that states that it'll not work.
Creating another temporary keychain only for password verification looks really ugly.
I ended up with map changed from `der => (Pkcs12, TempDir)` to `(der, hash(pass)) => (Pkcs12, TempDir)` using CommonCrypto bindings.

sfackler commented:
I'm talking about this use case in particular, not hash functions in general.

blackbeam commented:
> I'm talking about this use case in particular, not hash functions in general.

Code:
```rust
#[bench]
fn load_pkcs12(bencher: &mut tst::test::Bencher) {
    let buf = include_bytes!("../test/identity.p12");
    bencher.iter(|| Pkcs12::from_der(buf, "mypass").unwrap());
}
```

Default hasher:
```
test test::load_pkcs12 ... bench:       2,249 ns/iter (+/- 333)
```

xxHash:
```
test test::load_pkcs12 ... bench:       1,458 ns/iter (+/- 935)
```

sfackler commented:
Hmm, did you ever try using a single keychain for all of the identities? It seems like that'd simplify things quite a lot.

blackbeam commented:
I switched to a single keychain and also have any of password or der cache removed. Also looks like  it is ok to call SecPKCS12Import multiple times on the same der data with the same keychain being specified, however this does not stated clearly in the docs.
I've added `import_same_identity_multiple_times` test end [run it](https://travis-ci.org/blackbeam/rust-native-tls/builds/268885305?utm_source=github_status&utm_medium=notification) on every osx version available on travis.

sfackler commented:
Awesome, thanks!
